connectors/sync_job_runner.py

```diff
@@ -0,0 +1,175 @@
+        self.connector = connector
+        self.connector_id = self.connector.id
+        self.elastic_server = elastic_server
+        self.bulk_options = bulk_options
+        self._start_time = None
+        self.running = False
+
+    async def execute(self):
+        if self.running:
+            raise SyncJobRunningError(f"Sync job #{self.job_id} is already running.")
+
+        self.running = True
+        if not await self._claim_job():
+            logger.error(f"Unable to claim job #{self.job_id}")
+            raise JobClaimError
+
+        try:
+            self.data_provider = self.source_klass(self.sync_job.configuration)
+            if not await self.data_provider.changed():
+                logger.debug(
+                    f"No change in {self.sync_job.service_type} data provider, skipping..."
+                )
+                await self._sync_done(sync_status=JobStatus.COMPLETED, result={})
+                return
+
+            logger.debug(f"Validating configuration for {self.data_provider}")
+            await self.data_provider.validate_config()
+
+            logger.debug(
+                f"Syncing '{self.sync_job.service_type}' for connector '{self.connector_id}'"
+            )
+            logger.debug(f"Pinging the {self.source_klass} backend")
+            await self.data_provider.ping()
+
+            sync_rules_enabled = self.connector.features.sync_rules_enabled()
+            if sync_rules_enabled:
+                await self.sync_job.validate_filtering(validator=self.data_provider)
+
+            mappings = Mappings.default_text_fields_mappings(
+                is_connectors_index=True,
+            )
+
+            logger.debug("Preparing the content index")
+            await self.elastic_server.prepare_content_index(
+                self.sync_job.index_name, mappings=mappings
+            )
+
+            # allows the data provider to change the bulk options
+            bulk_options = self.bulk_options.copy()
+            self.data_provider.tweak_bulk_options(bulk_options)
+
+            result = await self.elastic_server.async_bulk(
+                self.sync_job.index_name,
+                self.prepare_docs(),
+                self.sync_job.pipeline,
+                filter_=self.sync_job.filtering,
+                sync_rules_enabled=sync_rules_enabled,
+                options=bulk_options,
+            )
+            sync_error = result.get("fetch_error")
+            sync_status = JobStatus.COMPLETED if sync_error is None else JobStatus.ERROR
+            await self._sync_done(
+                sync_status=sync_status, result=result, sync_error=sync_error
+            )
+        except asyncio.CancelledError:
+            await self._sync_done(sync_status=JobStatus.SUSPENDED, result={})
+        except Exception as e:
+            await self._sync_done(sync_status=JobStatus.ERROR, result={}, sync_error=e)
+        finally:
+            self._start_time = None
```

Comment:
Do we need to reset the `_start_time` here? (do we reuse a sync job runner instance)? Just curious.

Reply:
Not really, I copied from `connector.sync`. I will remove this.